File: ruby/lib/baseh/feistel.rb
# frozen_string_literal: true

require "openssl"

module Baseh
  # Balanced Feistel network with cycle walking, spec section 7.3.
  # HMAC-SHA-256 comes from OpenSSL; HMAC and SHA-256 are never implemented
  # by hand (section 7.5).
  module Feistel
    TAG = "BASEH-FEISTEL-V1".b
    MAX_WALKS = 1000

    module_function

    # ceil(log2(capacity)); capacity >= 2 so bits >= 1.
    def bit_length(capacity)
      (capacity - 1).bit_length
    end

    # Low n bits of the HMAC-SHA-256 digest: first ceil(n / 8) bytes read as
    # a big-endian integer and masked with 2^n - 1.
    def low_bits(digest, n)
      byte_count = (n + 7) / 8
      v = digest.byteslice(0, byte_count).unpack("C*").inject(0) { |acc, b| (acc << 8) | b }
      v & ((1 << n) - 1)
    end

    def to_be(value, byte_count)
      return "".b if byte_count.zero?

      bytes = Array.new(byte_count)
      v = value
      (byte_count - 1).downto(0) do |i|
        bytes[i] = v & 0xff
        v >>= 8
      end
      bytes.pack("C*")
    end

    # Normative round message, spec 7.3 step 4. In expandable mode the
    # generation's total code length L is mixed in as ASCII decimal plus a
    # 0x00 terminator after the profileId field (spec 7.3/19.4); fixed-mode
    # messages stay byte-for-byte unchanged.
    def round_message(profile_id, round, right, wr, length = nil)
      "".b
        .concat(TAG)
        .concat(0.chr(Encoding::BINARY))
        .concat(profile_id)
        .concat(0.chr(Encoding::BINARY))
        .concat(length.nil? ? "".b : "#{length}\x00".b)
        .concat(round.chr(Encoding::BINARY))
        .concat(to_be(right, (wr + 7) / 8))
    end

    def hmac(key_bytes, message)
      OpenSSL::HMAC.digest(OpenSSL::Digest.new("sha256"), key_bytes, message)
    end

    def run_rounds(left, right, profile_id, key_bytes, rounds, w0, w1, length)
      rounds.times do |i|
        even = i.even?
        wr = even ? w1 : w0
        wl = even ? w0 : w1
        f = low_bits(hmac(key_bytes, round_message(profile_id, i, right, wr, length)), wl)
        new_left = right
        new_right = left ^ f
        left = new_left
        right = new_right
      end
      [left, right]
    end

    def run_inverse(left, right, profile_id, key_bytes, rounds, w0, w1, length)
      (rounds - 1).downto(0) do |i|
        even = i.even?
        wr = even ? w1 : w0
        wl = even ? w0 : w1
        f = low_bits(hmac(key_bytes, round_message(profile_id, i, left, wr, length)), wl)
        prev_right = left
        prev_left = right ^ f
        left = prev_left
        right = prev_right
      end
      [left, right]
    end

    # Forward permutation with cycle walking. length is expandable mode only
    # (spec 7.3/19.4): the generation's total code length mixed into the round
    # message; omit it in fixed mode.
    def permute(value, capacity, profile_id:, key_bytes:, rounds:, length: nil)
      walk(value, capacity, profile_id, key_bytes, rounds, length) do |left, right, w0, w1|
        run_rounds(left, right, profile_id, key_bytes, rounds, w0, w1, length)
      end
    end

    # Inverse permutation with cycle walking.
    def inverse_permute(value, capacity, profile_id:, key_bytes:, rounds:, length: nil)
      walk(value, capacity, profile_id, key_bytes, rounds, length) do |left, right, w0, w1|
        run_inverse(left, right, profile_id, key_bytes, rounds, w0, w1, length)
      end
    end

    def walk(value, capacity, profile_id, key_bytes, rounds, length)
      bits = bit_length(capacity)
      w1 = bits / 2
      w0 = bits - w1
      v = value
      MAX_WALKS.times do
        left = v >> w1
        right = v & ((1 << w1) - 1)
        out_left, out_right = yield(left, right, w0, w1)
        combined = (out_left << w1) | out_right
        return combined if combined < capacity

        v = combined
      end
      raise BasehError.new(
        "PERMUTATION_FAILURE",
        "Feistel cycle walking exceeded 1000 iterations",
        safe_for_customer: false
      )
    end

    # Only permute and inverse_permute are public; the round machinery is
    # internal to the spec 7.3 permutation.
    class << self
      private :bit_length, :low_bits, :to_be, :round_message, :hmac,
              :run_rounds, :run_inverse, :walk
    end
  end
end

File: python/src/baseh/feistel.py
"""Feistel-v1 reversible permutation, spec section 7.3.

Message bytes, half widths, low-N-bits truncation and cycle walking are
implemented exactly as written in section 7.3. The vectors in
vectors/feistel-vectors.json are the ground truth.
"""

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass

from .errors import PERMUTATION_FAILURE, BasehError


@dataclass(frozen=True)
class FeistelKey:
    profile_id: str
    key_bytes: bytes
    rounds: int


_TAG = b"BASEH-FEISTEL-V1"
_MAX_WALKS = 1000


def _bit_length(capacity: int) -> int:
    return (capacity - 1).bit_length()


def _low_bits(digest: bytes, n: int) -> int:
    """Interpret the first ceil(n/8) digest bytes as big-endian, mask to n bits."""
    byte_count = (n + 7) // 8
    value = int.from_bytes(digest[:byte_count], "big")
    return value & ((1 << n) - 1)


def _round_message(profile_id: str, round_number: int, right: int, wr: int) -> bytes:
    pid_bytes = profile_id.encode("ascii")
    right_bytes = right.to_bytes((wr + 7) // 8, "big")
    return (
        _TAG
        + b"\x00"
        + pid_bytes
        + b"\x00"
        + bytes([round_number])
        + right_bytes
    )


def _run_rounds(left: int, right: int, key, w0: int, w1: int) -> tuple:
    for i in range(key.rounds):
        even = i % 2 == 0
        wr = w1 if even else w0
        wl = w0 if even else w1
        digest = hmac.new(
            key.key_bytes, _round_message(key.profile_id, i, right, wr), hashlib.sha256
        ).digest()
        f = _low_bits(digest, wl)
        left, right = right, left ^ f
    return left, right


def _run_inverse(left: int, right: int, key, w0: int, w1: int) -> tuple:
    for i in range(key.rounds - 1, -1, -1):
        even = i % 2 == 0
        wr = w1 if even else w0
        wl = w0 if even else w1
        digest = hmac.new(
            key.key_bytes, _round_message(key.profile_id, i, left, wr), hashlib.sha256
        ).digest()
        f = _low_bits(digest, wl)
        left, right = right ^ f, left
    return left, right


def _split(value: int, w1: int) -> tuple:
    return value >> w1, value & ((1 << w1) - 1)


def _combine(left: int, right: int, w1: int) -> int:
    return (left << w1) | right


def permute(value: int, capacity: int, key) -> int:
    """Forward permutation with cycle walking, spec 7.3 steps 1-7."""
    bits = _bit_length(capacity)
    w1 = bits // 2
    w0 = bits - w1
    v = value
    for _ in range(_MAX_WALKS):
        left, right = _split(v, w1)
        left, right = _run_rounds(left, right, key, w0, w1)
        out = _combine(left, right, w1)
        if out < capacity:
            return out
        v = out
    raise BasehError(
        PERMUTATION_FAILURE, "Feistel cycle walking exceeded 1000 iterations", False
    )


def inverse_permute(value: int, capacity: int, key) -> int:
    """Inverse permutation with identical cycle walking, spec 7.3."""
    bits = _bit_length(capacity)
    w1 = bits // 2
    w0 = bits - w1
    v = value
    for _ in range(_MAX_WALKS):
        left, right = _split(v, w1)
        left, right = _run_inverse(left, right, key, w0, w1)
        out = _combine(left, right, w1)
        if out < capacity:
            return out
        v = out
    raise BasehError(
        PERMUTATION_FAILURE, "Feistel cycle walking exceeded 1000 iterations", False
    )
